### PR TITLE
[DF] Re-use jitted lambdas for new RDF expressions with same logic

### PR DESCRIPTION
**Pull request by eguiraud**
This reduces the time required by jitting dramatically in many common scenarios, as we do not need a different template specialization of types that are expensive to instantiate for each Define or Filter expression.

Performance measurements available [here](https://indico.cern.ch/event/909884/)

### CODE REVIEW

File: tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx [hunk 1]
```diff
@@ -534,6 +535,15 @@ template <typename T>
 struct IsDeque_t<std::deque<T>> : std::true_type {};
 // clang-format on
 
+/// Return the static global map of experssions that have been jitted by RDF.
```
Comment (etejedor) [inline tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx:543]: `experssions` -> `expressions`

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx:543]: thanks, will fix

**Comment by eguiraud** [comment]
@etejedor thanks for taking a look!

1. correct. making concurrent usage of different RDFs from different threads safe is the topic of the recently created [ROOT-10694](https://sft.its.cern.ch/jira/browse/ROOT-10694)
2. avoiding re-jitting across computation graphs needs #5356 , but the general machinery is exercised by current tests (demonstrated by the current test failures) and tutorials (results I showed in the PPP were on df104, df105, df106). What kind of test would you like to see? (i.e. what should it test exactly?)

**Comment by etejedor** [comment]
I was thinking of something like this:

https://eguiraud.web.cern.ch/eguiraud/decks/20200416_reusing_rdf_jitted_lambdas_ppp/#/4

but booking all the histos before calling `GetValue` on one of them. I.e. some dedicated test with several RDFs and where reuse across RDFs clearly happens (of course other tests might exercise it to some degree already). Perhaps after merging #5356 , then.

**Comment by eguiraud** [comment]
But we can't _test_ that it definitely happens (and both tutorials and tests definitely exercise the feature already). This being a performance improvement rather than a bugfix, I would suggest we add a benchmark to rootbench rather than a test.

**Comment by etejedor** [comment]
> But we can't test that it definitely happens 

I agree, but at least you can know whether it breaks. Pushing the machinery with a dedicated test wouldn't hurt. But if you are happy with the coverage that the current tests already do, forget about my comment then.

File: tree/dataframe/src/RDFInterfaceUtils.cxx [hunk 2]
```diff
@@ -59,6 +61,140 @@ class RDataSource;
...
+using ROOT::Detail::RDF::ColumnNames_t;
+
+/// A string expression such as those passed to Filter and Define, digested to a standardized form
+struct ParsedExpression {
+   /// The string expression with the dummy variable names in fVarNames in place of the original column names
+   std::string fExpr;
+   /// The list of valid column names that were used in the original string expression.
+   /// Duplicates are removed and column aliases (created with Alias calls) are resolved.
+   ColumnNames_t fUsedCols;
+   /// The list of variable names used in fExpr, with same ordering and size as fVarUsedCols
```
Comment (hageboeck) [inline tree/dataframe/src/RDFInterfaceUtils.cxx:74]:
```suggestion
   /// The list of variable names used in fExpr, with same ordering and size as fUsedCols
```

File: tree/dataframe/src/RDFInterfaceUtils.cxx [hunk 3]
```diff
@@ -599,22 +735,20 @@ void BookFilterJit(const std::shared_ptr<RJittedFilter> &jittedFilter, void *pre
...
-   auto usedBranches = FindUsedColumnNames(expression, branches, customCols.GetNames(), dsColumns, aliasMap);
-   auto varNames = ReplaceDots(usedBranches);
-   auto dotlessExpr = std::string(expression);
-   const auto usedColTypes =
-      ColumnTypesAsString(usedBranches, varNames, aliasMap, tree, ds, dotlessExpr, customCols);
+   const auto parsedExpr =
+      ParseRDFExpression(std::string(expression), branches, customCols.GetNames(), dsColumns, aliasMap);
+   const auto exprVarTypes = GetColumnTypes(parsedExpr, tree, ds, customCols);
 
    TRegexp re("[^a-zA-Z0-9_]?return[^a-zA-Z0-9_]");
```
Comment (hageboeck) [inline tree/dataframe/src/RDFInterfaceUtils.cxx:796]: Didn't think about this much. What puzzles me, though, is that you want something that's not something before return, but that's optional. In the end, you don't care if there's anything before `return`.

Comment (hageboeck) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:796]: Note that I meant the line the github *doesn't show* here.

Comment (eguiraud) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:796]: we care if there is something before `return` that _is_ `a-zA-Z0-9_`: we don't want to match something like `"var_return"`

Comment (hageboeck) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:796]: That's what I meant:
```
TPRegexp re2("[^a-zA-Z0-9_]?return")
root [4] re2.Match("areturn")
(int) 1
```

Comment (eguiraud) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:796]: I see! Well that's a bug I introduced in February 2019 then :+1: Thanks for spotting it! I'll fix it in another PR if that's ok

Comment (eguiraud) [reply tree/dataframe/src/RDFInterfaceUtils.cxx:796]: Fixed in #5487

**Pull request merged**